client/boot/common.js

```diff
@@ -36,41 +36,34 @@ const switchUserLocale = ( currentUser, reduxStore ) => {
 
 const setupContextMiddleware = reduxStore => {
 	page( '*', ( context, next ) => {
-		const parsed = url.parse( location.href, true );
-
-		// Decode the pathname by default (now disabled in page.js)
-		context.pathname = decodeURIComponent( context.pathname );
-
-		context.store = reduxStore;
-
-		// Break routing and do full load for logout link in /me
-		if ( context.pathname === '/wp-login.php' ) {
-			window.location.href = context.path;
-			return;
-		}
-
-		// set `context.query`
-		const querystringStart = context.canonicalPath.indexOf( '?' );
-
-		if ( querystringStart !== -1 ) {
-			context.query = qs.parse( context.canonicalPath.substring( querystringStart + 1 ) );
-		} else {
-			context.query = {};
-		}
-
+		// page.js url parsing is broken so we had to disable it with `decodeURLComponents: false`
+		const parsed = url.parse( context.canonicalPath, true );
+		context.pathname = parsed.pathname;
 		context.prevPath = parsed.path === context.path ? false : parsed.path;
+		context.path = parsed.path;
+		context.query = parsed.query;
 
+		context.hashstring = ( parsed.hash && parsed.hash.substring( 1 ) ) || '';
 		// set `context.hash` (we have to parse manually)
-		if ( parsed.hash && parsed.hash.length > 1 ) {
+		if ( context.hashstring ) {
 			try {
-				context.hash = qs.parse( parsed.hash.substring( 1 ) );
+				context.hash = qs.parse( context.hashstring );
```

Comment:
Nitpick: Do we really need the `try` here?
It seems this throws only in a specific case not really related to parsing: https://github.com/ljharb/qs/blob/master/lib/parse.js#L142

Reply:
It doesn't look like we need it 👍